% \VignetteIndexEntry{RPA}
%The above line is needed to remove a warning in R CMD check
\documentclass[12pt]{article}

\usepackage{amsmath,amssymb,amsfonts}
\usepackage{hyperref}
%\usepackage[authoryear,round]{natbib}
\usepackage{Sweave}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\def\muj{\mathbf{\mu}_j}
\def\mujs{\{\mathbf{\mu}_j\}}

\def\s{\{\mathbf{s}\}}
\def\st{\mathbf{s}^t}
\def\sj{\mathbf{s}_j}
\def\sij{s_{ij}}
\def\sjs{\mathbf{s}_{j=1}^J}
\def\sjt{\mathbf{s}_j^t}
\def\stj{s_{tj}}
\def\scj{s_{cj}}
\def\stjs{\{s_{tj}\}_t}
\def\sjI{\mathbf{s}_j^1}
\def\sjT{\mathbf{s}_j^T}

\def\m{\mathbf{m}}
\def\mj{\mathbf{m}_j}
\def\mjs{\{\mathbf{m}_j\}}
\def\mtj{m_{tj}}

\def\gj{\{\mathbf{g}_j\}}
\def\gi{g_i}
\def\gt{g_t}
\def\gc{g_c}
\def\g{\mathbf{g}}

\def\d{\mathbf{d}}
\def\dt{d_t}
\def\djt{\mathbf{d}_j^t}
\def\djr{\mathbf{s}_{j}}

\def\TauSq{\boldsymbol{\tau}^2}
\def\tauj{\mathbf{\tau}_j}
\def\taus{\{\tau_j^2\}}
\def\taujSq{\mathbf{\tau}_j^2}
\def\tauOneSq{\mathbf{\tau}_1^2}
\def\tauPSq{\mathbf{\tau}_J^2}

\def\epsilonj{\varepsilon_j}
\def\epsilonij{\mathbf{\varepsilon}_{ij}}
\def\epsiloncj{\varepsilon_{cj}}
\def\epsilontj{\varepsilon_{tj}}
\def\epsilontjs{\{\varepsilon_{tj}\}_t}
\def\epsiloncjs{\{\varepsilon_{cj}\}_j}
\def\Epsilonj{\boldsymbol{\varepsilon}_j}
\def\Epsiloncj{\boldsymbol{\varepsilon}_{cj}}

\def\alphaj{\alpha_j}
\def\betaj{\beta_j}
\def\alphahatj{\hat{\alpha}_j}
\def\betahatj{\hat{\beta}_j}

\def\zi{z_i}
\def\j{\mathbf{j}}
\def\sigmaj{\sigma_j}
\def\lambdaj{\lambda_j}

\author{Leo Lahti\footnote{http://www.iki.fi/Leo.Lahti}\\University of Helsinki\\leo.lahti@iki.fi}
  
\begin{document} \title{Robust Probabilistic Averaging (RPA)}

\maketitle 

\section{Introduction}
 
\Rpackage{RPA} (Robust Probabilistic
Averaging)\footnote{\url{http://bioconductor.org/packages/release/bioc/html/RPA.html}}
provides tools for probe reliability analysis and gene expression
preprocessing for (Affymetrix and other) short oligonucleotide arrays,
and more generally to summarize normally distributed multivariate
observations that target the same object with varying degrees of
reliability.  RPA provides explicit data-driven estimates of probe
performance, i.e. affinity and probe-specific noise level
\cite{Lahti11}, and a fully scalable approach for preprocessing
arbitrarily large collections of CEL files. It has been shown to
outperform the standard RMA algorithm \cite{Irizarry03rma} in various benchmarking tests \cite{Lahti11, Lahti12} and in contrast to fRMA \cite{McCall2011} it
is readily applicable to all Affymetrix platforms.

\section{Usage examples}

For usage examples, see https://github.com/microbiome/microbiome/wiki/RPA

\section{Methodology}

For details, see \cite{Lahti11, Lahti12}.

\section{Citing RPA}

Please cite \cite{Lahti11, Lahti12}.

\section{Details}

This document was written using:

<<details>>=
sessionInfo()
@

%\bibliographystyle{abbrv}
%\bibliography{RPA}

\begin{thebibliography}{10}

\bibitem{Irizarry03rma}
R.~A. Irizarry, B.~M. Bolstad, F.~Collin, L.~M. Cope, B.~Hobbs, and T.~P.
  Speed.
\newblock {Summaries of Affymetrix GeneChip probe level data}.
\newblock {\em Nucl. Acids Res.}, 31(4):e15, 2003.

\bibitem{Lahti11}
L.~Lahti, L.~L. Elo, T.~Aittokallio, and S.~Kaski.
\newblock Probabilistic analysis of probe reliability in differential gene
  expression studies with short oligonucleotide arrays.
\newblock {\em {IEEE/ACM Transactions on Computational Biology and
  Bioinformatics}}, 8(1):217--225, 2011.

\bibitem{Lahti12}
L.~Lahti, A. Torrente, L.~L. Elo, A. Brazma, J. Rung.
\newblock A fully scalable online-preprocessing algorithm for short oligonucleotide microarray atlases
\newblock In review. arXiv preprint available at: http://arxiv.org/abs/1212.5932

\bibitem{McCall2011}
MN McCall and RA Irizarry.
\newblock Thawing Frozen Robust Multi-array Analysis (fRMA)
\newblock {\em BMC Bioinformatics} 12:369, 2011.

\end{thebibliography}


\end{document}

